--- a/gnome-session/gsm-manager.c
+++ b/gnome-session/gsm-manager.c
@@ -835,7 +835,10 @@
         GsmManagerPrivate *priv = gsm_manager_get_instance_private (data->manager);
         gboolean ret;
         GError  *error;
-
+        if (g_strcmp0 ("org.gnome.Shell.desktop", gsm_client_peek_app_id (client)) == 0) {
+                g_debug ("GsmManager: skipping EndSession for gnome-shell");
+                return FALSE;
+        }
         error = NULL;
         ret = gsm_client_end_session (client, data->flags, &error);
         if (! ret) {
@@ -1024,7 +1027,10 @@
         GsmManagerPrivate *priv = gsm_manager_get_instance_private (data->manager);
         gboolean ret;
         GError  *error;
-
+        if (g_strcmp0 ("org.gnome.Shell.desktop", gsm_client_peek_app_id (client)) == 0) {
+                g_debug ("GsmManager: skipping QueryEndSession for gnome-shell");
+                return FALSE;
+        }	
         error = NULL;
         ret = gsm_client_query_end_session (client, data->flags, &error);
         if (! ret) {
